Background bulk export from clipboard, Refs #9998

sbreker opened the pull request.

sbreker committed:
Create XML Export background job, Refs #9998

Created a new background job that wraps the existing XML export CLI
utility to provide asynchronous export capability to the WebUI. XML output
is downloadable from the Job Management page as a zip file.

Modified the existing CSV export job to allow an option to include
descendant records in the CSV export results.

  Changes to be committed:
	modified:   config/gearman.yml
	modified:   lib/job/arInformationObjectCsvExportJob.php
	new file:   lib/job/arXmlExportJob.php

sbreker committed:
New Clipboard CSV/XML Export Form, Refs #9998

This commit includes:
- a new form accessed via the clipboard page for setting options for
CSV and XML Exporting from the clipboard.
- a new js file that controls form behaviour and validation.
- modifications to the Clipboard page template to include an 'Export'
button and removal of the old 'Export CSV' link.

Note that the new form includes settings for options that will be
implemented in upcoming tickets.

  Changes to be committed:
	new file:   apps/qubit/modules/object/actions/exportAction.class.php
	modified:   apps/qubit/modules/object/config/view.yml
	new file:   apps/qubit/modules/object/templates/exportResultsSuccess.php
	new file:   apps/qubit/modules/object/templates/exportSuccess.php
	modified:   apps/qubit/modules/user/templates/clipboardSuccess.php
	new file:   js/exportOptions.js

sbreker committed:
Clipboard menu items by class name, Refs #9998

This commit includes changes to the Clipboard menu dropdown to display
a count of items in the clipboard by class name.

QubitClipboard has been changed in terms of how the clipboard items are
serialized - instead of a flat array of slugs going into sfStorage, the
array is now one level deeper so that slugs are grouped by class name.
This was done to ensure that the class name was determined when an item
is added to storage, as opposed to figuring out the class name of all
objects in the clipboard when viewing the menu.

 Changes to be committed:
	modified:   apps/qubit/modules/menu/actions/clipboardMenuComponent.class.php
	modified:   apps/qubit/modules/menu/templates/_clipboardMenu.php
	modified:   apps/qubit/modules/user/actions/clipboardToggleSlugAction.class.php
	modified:   js/clipboard.js
	modified:   lib/QubitClipboard.class.php

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,159 @@
+    // We will use this later to redirect users back to the importSelect page
+    if (isset($this->getRoute()->resource))
+    {
+      $exportRoute = array($this->getRoute()->resource, 'module' => 'object', 'action' => 'importSelect', 'type' => $exportType);
```
Shouldn't it be to the export action? If so, I think you don't need this check nor the redirect in the following exception catch.

sbreker replied:
Good catch - cut and paste error on my part.  Will adjust this.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,159 @@
+    catch (sfException $e)
+    {
+      $this->context->user->setFlash('error', $e->getMessage());
+      $this->redirect($exportRoute);
```
If I was right before, leaving the user in the same route with the flash alert should be enough. You could use `return sfView::SUCCESS;`

sbreker replied:
Changing this to `return sfView::SUCCESS` does not seem to work. There are two undesirable effects when I use that:
- I want to have the 'type' specified on the URL to pre-set one of the dropdowns.
- If execution is in this catch block we have an error.  On error, I want to do a http GET on this form and re-load with the 'type' set.  Returning sfView::SUCCESS appears to do a POST and makes it look like the export completed successfully. I have tried 'sfView::ERROR' but still see the form being POSTed.

jraddaoui replied:
It could be stored in `$this->type` making it available in the template instead of sending it in the request. I think we could remove the checks to build the URL and avoid the redirect that way.

MikeFE replied:
can you use $this->var = blah when using return sfView::SUCCESS? I can't seem to get it to work!

jraddaoui replied:
It works for me. Could it be that it was not set in the URL? It's checking it to populate the var.

And that's other thing I don't like very much, we're calling the same thing three different names: format, type and exportType in the same action, which creates a lot of confusion and unneeded code.

sbreker replied:
I think we may have been misunderstanding each other.  I have changed this and have code working using the `return sfView::SUCCESS`.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,159 @@
+
+        $this->doBackgroundExport($request);
+
+        $this->setTemplate('exportResults');
```
I'd redirect to the previous page, with the flash alert should be enough here too. That way we don't need the exportResults template, which redirects you to the export page again. I'm not completely sure that the following will work, but it should be something like `$this->redirect($request->getReferer());`

jraddaoui replied:
You may need to check that the "referer" exists and if it doesn't I'd let it reach the default template

sbreker replied:
Sounds good - I've made this change.  The flash alert does still show up with a link to the job management page.  Not sure what you mean by 'default template' - if (null == getReferer()) then I redirect to the export page: `$this->redirect(array('module' => 'object', 'action' => 'export'));`

jraddaoui replied:
If it reaches the end of the execute method without a redirect or a return it will load the default template, the same as doing `return sfView::SUCCESS` at the end

jraddaoui commented on apps/qubit/modules/object/templates/exportResultsSuccess.php:
```diff
@@ -0,0 +1,7 @@
+  <ul>
+    <li><?php echo link_to(__('Back'), array('module' => 'object', 'action' => 'export'), array('class' => 'c-btn')) ?></li>
+  </ul>
+</section>
```
I'd remove this template if you could redirect to the previous page where the export was initiated

sbreker replied:
Removed.

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -0,0 +1,103 @@
+            <select name="format">
+              <option value="xml" selected="selected"><?php echo __('XML') ?></option>
+              <option value="csv"><?php echo __('CSV') ?></option>
+            </select>
```
An inline if in the option elements may be better in here to avoid repeating the whole select block.

sbreker replied:
Done.

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -0,0 +1,103 @@
+                <?php echo $form->levels
+                  ->label(__('Levels of description'))
+                  ->help(__('Select the levels of description to be included in the export. If no levels are selected, the export will fail. You can use the control (Mac ⌘) and/or shift keys to multi-select values from the Levels of description menu. Descriptions that are descendants of levels not included in the export will also be excluded.'))
+                  ->renderRow(array('class' => 'form-autocomplete')) ?>
```
Not sure if the "form-autocomplete" class is needed here

sbreker replied:
Tested and you are correct. I have removed the array from renderRow().

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -0,0 +1,103 @@
+                  ->renderRow(array('class' => 'form-autocomplete')) ?>
+              </div>
+              <br />
+              <?php $taxonomy = QubitTaxonomy::getById(QubitTaxonomy::LEVEL_OF_DESCRIPTION_ID) ?>
```
Is this intended? I think it's not being used

sbreker replied:
Removed.

jraddaoui commented on apps/qubit/modules/object/templates/exportSuccess.php:
```diff
@@ -0,0 +1,103 @@
+    <section class="actions">
+      <ul>
+        <li><input class="c-btn c-btn-submit" type="submit" id="exportSubmit" value="<?php echo __('Export') ?>"/></li>
+        <li><?php echo link_to(__('Cancel'), array('module' => 'user', 'action' => 'clipboard'), array('class' => 'c-btn')) ?></li>
```
I'd use the "referer" in here too, I think it will keep the params in the clipboard page if there are any. And if will be ready for the future if we link to here from other places. The request can be accessed from the template with `$sf_request`

jraddaoui replied:
You may need to check that the "referer" exists and if it doesn't I'd redirect to the clipboard page (until this can be reached from other places)

sbreker replied:
Yes, I can see the value in this in case we add additional links to this form. Change made.

jraddaoui commented on lib/job/arInformationObjectCsvExportJob.php:
```diff
@@ -67,7 +68,7 @@ public function runJob($parameters)
     // Export CSV to temp directory
     $this->info($this->i18n->__('Starting export to %1.', array('%1' => $tempPath)));
-    $itemsExported = $this->exportResults($tempPath);
+    $itemsExported = $this->exportResults($tempPath, $this->params);
```
`$this->params` should be available in there without sending it

sbreker replied:
Done.

jraddaoui commented on lib/job/arInformationObjectCsvExportJob.php:
```diff
@@ -121,6 +122,17 @@ protected function exportResults($path)
         $writer->exportResource($resource);
 
+        // export descendants if configured
+        if (true != $this->params['current-level-only'])
```
`if (!$this->params['current-level-only'])` looks better to me

sbreker replied:
Completed.

jraddaoui commented on lib/job/arXmlExportJob.php:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    return $success;
+  }
```
The createZipForDownload function could be moved to arBaseJob and removed from both CSV and XML export jobs

sbreker replied:
Completed.

jraddaoui commented on lib/job/arXmlExportJob.php:
```diff
@@ -0,0 +1,190 @@
+   */
+  protected function exportResults($path, $params)
+  {
+    $exitCode= 0;
```
Missing space

sbreker replied:
Change made.

jraddaoui commented:
Hi @sbreker, nice work!

I've added a few comments along the changes and, in general, I wonder if this should be working with actors and repositories. I could not see any way to add those to the clipboard and, in both jobs, it looks like all kind of slugs can be sent but only the information object index is queried.

Also, I'd appreciate if you could update the pull request title and description with the commit title and description you will use in the merge. Thanks!

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,159 @@
+        QubitJob::runJob('arXmlExportJob', $options);
+      }
+
+      // Let user know import has started
```
Export ;)

sbreker replied:
Fixed.  :)

sbreker commented:
The work to add export functionality for the additional entity types (repos & actors) has been broken out into other tickets.  This task completes the WebUI changes and creates the XML export background job.

sbreker committed:
Background bulk export from clipboard, Refs #9998

- modified export route used when redirecting to form on error.
- change form to use referer url when redirecting back to originating page.
- condensed code for the 'type' dropdown to avoid repeated code.
- removed exportResultsSuccess template in favour of redirecting to main
form page after export is initiated.
- removed class 'form-autocomplete' from levels selection widget.
- changed export form cancel button to use referer.
- moved createZipForDownload() method to arBaseJob to reduce code duplication.
- additional minor and cosmetic changes.

	modified:   apps/qubit/modules/menu/templates/_clipboardMenu.php
	modified:   apps/qubit/modules/object/actions/exportAction.class.php
	deleted:    apps/qubit/modules/object/templates/exportResultsSuccess.php
	modified:   apps/qubit/modules/object/templates/exportSuccess.php
	modified:   js/exportOptions.js
	modified:   lib/job/arBaseJob.class.php
	modified:   lib/job/arInformationObjectCsvExportJob.php
	modified:   lib/job/arXmlExportJob.php

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,166 @@
+
+      $this->title = $this->context->i18n->__('Export');
+    }
+  }
```
I'd do all this else block at the beginning of the function, outside the method condition, that way it will be all ready even if the form is not valid in a POST.

sbreker replied:
Change made.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,166 @@
+        else
+        {
+          $this->redirect(array('module' => 'object', 'action' => 'export'));
+        }
```
If you place the next else work at the beginning of the function like I say in the previous comment, this else and redirect to the same page shouldn't be needed.

sbreker replied:
Change made.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,166 @@
+    else
+    {
+      $exportRoute = array('module' => 'object', 'action' => 'export', 'type' => $exportType);
+    }
```
Actually, this looks like it comes from the import page, where you can import directly to a resource, but it should not be needed in the export page (AFAIK it can't be reached from a resource). And, as I said in other comments, if you move the `$this->limit` population outside the method check in the execute function, you should be able to access `$type` in the template in all cases.

sbreker replied:
I've removed this code.

jraddaoui commented:
Thanks @sbreker, just a few more comments about the redirect and the request method check that I think it will make it a little cleaner.

A detailed description in the merged commit (and the PR) will help to understand the feature and, for following PRs, if you keep the commit history intact while it's being reviewed, it makes it easier for the reviewer, but that may be my personal opinion ;)

Thanks!!

MikeFE commented on apps/qubit/modules/menu/templates/_clipboardMenu.php:
```diff
@@ -15,13 +15,10 @@
-        <?php endif; ?>
+      <span id="count-block" data-information-object-label="<?php echo sfConfig::get('app_ui_label_informationobject') ?>" data-actor-object-label="<?php echo sfConfig::get('app_ui_label_actor') ?>" data-repository-object-label="<?php echo sfConfig::get('app_ui_label_repository') ?>">
+        <?php foreach ($countByType as $objectType => $countType): ?>
+          <?php echo $objectTypes[$objectType] . ' count: ' . $countType ?><br />
```
don't forget __() for the 'count:' string

jraddaoui replied:
Thanks Mike ;)

sevein replied:
You have to use replacements, remember!

`<?php echo __('%1% count: %2%', array('%1%' => $objectTypes[$objectType], '%2%' => $countType)); ?>`

The translator will see `%1% count: %2%` which is going to be really hard to translate without contextual markers but I don't think they're supported in Symfony. Django does: https://docs.djangoproject.com/en/1.10/topics/i18n/translation/#contextual-markers.

MikeFE commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,166 @@
+      // Let user know export has started
+      sfContext::getInstance()->getConfiguration()->loadHelpers(array('Url'));
+      $jobManageUrl = url_for(array('module' => 'jobs', 'action' => 'browse'));
+      $message = '<strong>Export of descriptions initiated.</strong> Check <a href="'. $jobManageUrl . '">job management</a> page to download the results when it has completed.';
```
do we need i18n here too?

jraddaoui replied:
That's wrapped after, which I have no idea if it works ...

sevein replied:
That's not going to work, the i18n extractor finds the strings statically. You need to put the string literal inside `__()` AFAIK.

If you need replacements, use:

``` php
__('My name is %1%', array('%1%' => $name));
```

We try not to include HTML tags in our i18n strings, e.g.:

``` php
__('Export of descriptions initiated. %1%Check job management%2% page to download the results when it has completed.', array(
    '%1%' => sprintf('<a href="%s">', url_for(...)),
    '%2' => '</a>'));
```

I think the idea is to avoid having the translator writing HTML, but I'm not sure if AtoM is consistently doing that, probably not.

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,158 @@
+    if (isset($request->type))
+    {
+      $this->type = $request->type;
+    }
```
I think a default value is needed in `$this->type` for GET requests, as it's being used in the template in the type field. I'd replace this if with the way it's being done in the doBackgroundExport function:

`$this->type = $request->getParameter('type', 'csv');`

jraddaoui commented on apps/qubit/modules/object/actions/exportAction.class.php:
```diff
@@ -0,0 +1,158 @@
+  protected function doBackgroundExport($request)
+  {
+    // Export type, CSV or XML?
+    $this->type = $request->getParameter('type', 'csv');
```
It could be removed from here after it's added in the execute function.

jraddaoui commented:
👍 Thanks @sbreker !!
